### PR TITLE
added global exception handler

### PR DESCRIPTION
**Pull request by Oleksandr-Tymoshenko**

### CODE REVIEW

**Review by mmosii — APPROVED** [review]
Great job!

File: pom.xml [hunk 1]
```diff
@@ -59,6 +59,10 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
+            <artifactId>hibernate-validator</artifactId>
+        </dependency>
```
Comment (NazarSmith228) [inline pom.xml]: it is better to use `spring-boot-starter-validation`

File: src/main/java/book/store/onlinebookstore/controller/BookController.java [hunk 2]
```diff
@@ -39,12 +40,13 @@ public List<BookDto> search(BookSearchParameters searchParameters) {
...
 
     @PostMapping
-    public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
 
     @PutMapping("/{id}")
-    public BookDto updateById(@PathVariable Long id, @RequestBody CreateBookRequestDto bookDto) {
+    public BookDto updateById(@PathVariable Long id,
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/controller/BookController.java:48]: just FYI

you may also validate such things as @PathVariable, @RequestParams etc. It looks like:
```
@PathVariable @Positive Long id
```

In order to be able to perform such validation you need to annotate your controller class with @Validated annotation

again, this is just FYI - apply it or not :)

File: src/main/java/book/store/onlinebookstore/dto/BookDto.java [hunk 3]
```diff
@@ -3,6 +3,6 @@
 import java.math.BigDecimal;
 
 public record BookDto(Long id, String title, String author, String isbn,
-                       BigDecimal price, String description, String coverImage) {
+                      BigDecimal price, String description, String coverImage) {
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/dto/BookDto.java]: usually fields of records are formatted as follows:
```
public record BookDto(Long id, 
                      String title, 
                      String author, 
                      String isbn,
                      BigDecimal price, 
                      String description, 
                      String coverImage) {
```

File: src/main/java/book/store/onlinebookstore/dto/CreateBookRequestDto.java [hunk 4]
```diff
@@ -1,8 +1,15 @@
...
+import jakarta.validation.constraints.Pattern;
 import java.math.BigDecimal;
 
-public record CreateBookRequestDto(String title, String author, String isbn,
-                                   BigDecimal price, String description, String coverImage) {
+public record CreateBookRequestDto(
+        @NotNull String title,
+        @NotNull String author,
+        @NotNull @Pattern(regexp = "^(?=(?:\\D*\\d){10}(?:(?:\\D*\\d){3})?$)[\\d-]+$") String isbn,
+        @NotNull @Min(0) BigDecimal price,
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/dto/CreateBookRequestDto.java]: is it really necessary to use `@Min(0)`?

could you use `@Positive`?

File: src/main/java/book/store/onlinebookstore/dto/CreateBookRequestDto.java [hunk 5]
```diff
@@ -1,8 +1,15 @@
...
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Pattern;
 import java.math.BigDecimal;
 
-public record CreateBookRequestDto(String title, String author, String isbn,
-                                   BigDecimal price, String description, String coverImage) {
+public record CreateBookRequestDto(
+        @NotNull String title,
+        @NotNull String author,
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/dto/CreateBookRequestDto.java]: it is better to use `@NotBlank` with strings

File: src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers,
+            HttpStatusCode status,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("time: ", LocalDateTime.now());
+        body.put("status: ", HttpStatus.BAD_REQUEST);
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java]: time is pretty useless in response body - it does not bring any value to the client

status is obvious as well - no sense to have it in body

File: src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java:21]: do you have any other exceptions except for `MethodArgumentNotValidException`?

does it make sense to handle them as well?

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/book/store/onlinebookstore/controller/BookController.java (hunk 2 above)
Comment (Oleksandr-Tymoshenko) [reply src/main/java/book/store/onlinebookstore/controller/BookController.java:48]: I`ve added it in the next task

File: src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+                .toList();
+        body.put("errors: ", errors);
+        return new ResponseEntity<>(body, headers, status);
+    }
+
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<String> handleAllExceptions(Exception ex) {
+        return new ResponseEntity<>("An unexpected error occurred" + ex.getMessage(),
+                HttpStatus.INTERNAL_SERVER_ERROR);
+    }
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java]: it is a bad idea :)

better to handle specific exceptions

you can remove this handler

File: src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    @ExceptionHandler(EntityNotFoundException.class)
+    public ResponseEntity<String> handleNullPointerExceptions(EntityNotFoundException ex) {
+        return new ResponseEntity<>(ex.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java]:
```suggestion
    public ResponseEntity<String> handleEntityNotFound(EntityNotFoundException ex) {
        return new ResponseEntity<>(ex.getMessage(), HttpStatus.NOT_FOUND);

```

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Just a few remarks left

**Review by NazarSmith228 — APPROVED** [review]
LGTM

**Pull request merged**